plutus-tx/test/Plugin/Spec.hs

```diff
@@ -392,10 +403,157 @@ evenMutual = plc @"evenMutual" (
+
+toMyMonoData2 :: Either (Int, Int) Int -> MyMonoData2
+toMyMonoData2 (Left  (i, j)) = Mono21 i j
+toMyMonoData2 (Right i     ) = Mono22 i
+
+readMyMonoData2 :: CompiledCode MyMonoData2 -> MyMonoData2
+readMyMonoData2 = readCompiledCode toMyMonoData2
+
+myMono21 :: MyMonoData2
+myMono21 = Mono21 0 1
+
+myMono22 :: MyMonoData2
+myMono22 = Mono22 2
+
+compiledMyMono21 :: CompiledCode MyMonoData2
+compiledMyMono21 = plc @"compiledMyMono21" (Mono21 0 1)
+
+compiledMyMono22 :: CompiledCode MyMonoData2
+compiledMyMono22 = plc @"compiledMyMono22" (Mono22 2)
+
+readMyMonoData2s :: CompiledCode [MyMonoData2] -> [MyMonoData2]
+readMyMonoData2s = readCompiledCode (Prelude.map toMyMonoData2 . PLC.unPlcList)
+
+myMonoData2s :: [MyMonoData2]
+myMonoData2s = [Mono22 2, Mono21 0 1, Mono21 4 3, Mono22 8, Mono21 6 7]
+
+compiledMyMonoData2s :: CompiledCode [MyMonoData2]
+compiledMyMonoData2s = plc @"compiledMyMonoData2s" (
+        [Mono22 2, Mono21 0 1, Mono21 4 3, Mono22 8, Mono21 6 7]
+    )
+
+data MyMonoData3 = Mono31 Int | Mono32 Int
+    deriving (Show, Eq)
+
+toMyMonoData3 :: Either Int Int -> MyMonoData3
+toMyMonoData3 (Left  i) = Mono31 i
+toMyMonoData3 (Right i) = Mono32 i
+
+readMyMonoData3 :: CompiledCode MyMonoData3 -> MyMonoData3
+readMyMonoData3 = readCompiledCode toMyMonoData3
+
+myMono31 :: MyMonoData3
+myMono31 = Mono31 0
+
+myMono32 :: MyMonoData3
+myMono32 = Mono32 2
+
+compiledMyMono31 :: CompiledCode MyMonoData3
+compiledMyMono31 = plc @"compiledMyMono31" (Mono31 0)
+
+compiledMyMono32 :: CompiledCode MyMonoData3
+compiledMyMono32 = plc @"compiledMyMono32" (Mono32 2)
+
+readMyMonoData3s :: CompiledCode [MyMonoData3] -> [MyMonoData3]
+readMyMonoData3s = readCompiledCode (Prelude.map toMyMonoData3 . PLC.unPlcList)
+
+myMonoData3s :: [MyMonoData3]
+myMonoData3s = [Mono31 2, Mono31 0, Mono31 4, Mono32 8, Mono31 6]
+
+compiledMyMonoData3s :: CompiledCode [MyMonoData3]
+compiledMyMonoData3s = plc @"compiledMyMonoData3s" (
+        [Mono31 2, Mono31 0, Mono31 4, Mono32 8, Mono31 6]
+    )
+
+-- data MyMonoData = Mono1 Int Int | Mono2 Int | Mono3 Int
+
+-- This gives this error:
+-- >      error: undefined reference to 'PluginziSpec_MyNewtype_closure'
+-- >     |
+-- > 427 | readMyNewtype = readCompiledCode MyNewtype
+-- readMyNewtype :: CompiledCode MyNewtype -> MyNewtype
+-- readMyNewtype = readCompiledCode MyNewtype
+
+myNewtype42 :: MyNewtype
+myNewtype42 = MyNewtype 42
+
+compiledMyNewtype42 :: CompiledCode MyNewtype
+compiledMyNewtype42 = plc @"myNewtype42" (MyNewtype 42)
+
+readDyns :: TestTree
+readDyns = testGroup "reads" [
+    testCase "readMyMonoRecord" $ readMyMonoRecord compiledMyMonoRecord01 @?= myMonoRecord01
+    , testCase "readMyMonoRecords" $ readMyMonoRecords compiledMyMonoRecords @?= myMonoRecords
+    -- All of these result in "Can't read compiled code".
+    -- The reason for this is that a constructor carrying two arguments does not compile to
+    -- a full product type when there are several constructors.
```

Comment:
Ah yes. So the sums-of-products representation is isomorphic but not the same, annoying.